client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java

```diff
@@ -136,6 +136,11 @@ public static int celebornShuffleId(
     }
   }
 
+  public static int getMapAttemptNumber(TaskContext context) {
+    assert (context.stageAttemptNumber() < (1 << 15) && context.attemptNumber() < (1 << 16));
+    return (context.stageAttemptNumber() << 16) | context.attemptNumber();
+  }
+
```

Comment:
(When `celeborn.client.spark.fetch.throwsFetchFailure` is not enabled)
We should make this assumption as it is not enforced by spark.
I do agree this is a ridiculously large number, but since spark does not put a bound on how large users can set these configs to; it can end up breaking (For life of me I cant image such a high value though ! )

What I would suggest is to first enforce it in spark - since spark 4.0 is going on, this might be a good time to add this enforcement: and if it is accepted by the spark community, we can leverage it here.

Reply:
IMO, enabling `throwsFetchFailure`  doesn't solve this problem. Image a barrier stage which reads data from an external datasource and does shuffle write.  If one of its tasks(assuming it is "map 10") encounters an exception while reading data, the barrier stage will be resubmitted. celeborn will reuse the shuffleId if the stage is not marked as INDETERMINATE. In this case, we will get 2 success MapperEnd(assuming they are "stage 0 stageAttempt 0 map 10 taskAttempt 0" and "stage 0 stageAttempt 1 map 10 taskAttempt 0") and we cannot distinguish them for now.

I will try to communicate with the spark community to see if we can add enforcement for the max attempt limit, such as changing the data type to Short, although it might not be easy to succeed. For now, we can only check the max attempt limit when initializing the SparkShuffleManager.

Reply:
When `throwsFetchFailure` is enabled, `SparkUtils.celebornShuffleId` considers the stage attempt id.
Did you try reproducing the issue with the flag enabled ?

Reply:
Re working with spark community - no, datatype change wont be allowed - agree.
I was suggesting that you propose adding a constraint on the value [here](https://github.com/apache/spark/blob/f2dd0b3338a6937bbfbea6cd5fffb2bf9992a1f3/core/src/main/scala/org/apache/spark/internal/config/package.scala#L2426) (checkValue); not change datatype.

While I will review it, this will require consensus in the community.

But before we go down that path - it would be better to validate if the issue is relevant with `throwsFetchFailure` enabled.

Reply:
I reproducing the issue at https://github.com/apache/celeborn/pull/2611, without that PR, the barrier stage will reuse shuffleId and we will get 2 success MapperEnd, "stage 1 stageAttempt 0 map 0 taskAttempt 0" and "stage 1 stageAttempt 1 map 0 taskAttempt 0"

Reply:
I reproducing the issue at https://github.com/apache/celeborn/pull/2611, without that PR, the barrier stage will reuse shuffleId and we will get 2 success MapperEnd, "stage 1 stageAttempt 0 map 0 taskAttempt 0" and "stage 1 stageAttempt 1 map 0 taskAttempt 0"

Reply:
To reproduce this problem, the following conditions must be met:
1. Resubmit stage and reuse celeborn shuffle id
2. There are at least two valid MapperEnds with same shuffleId, mapId, and taskAttemtNumber

Normally, these two conditions cannot be met at the same time. Because if stage is INDETERMINATE, celeborn won't reuse celeborn shuffle id, if stage is non-INDETERMINATE, DAGScheduler will only submit missing tasks which means no duplicate task will be submit. But resubmitting a non-INDETERMINATE barrier stage can meet there two conditions at the same time. 

Even worse, when the barrier stage is UNORDERED, shuffle read will find two data blocks with the same shuffleId, mapId, taskAttemptNumber, and batchId. According to the deduplication rule, celeborn will only read one of the batches and think that the other is just a duplicate batch, but in fact the two batches may be different, resulting in data correctness issue.

Reply:
Do we have a reproducible test where the issue is surfaced ? That is, some test which fails and is expected to work ? We can explore what the fix is once we have a reproducible test.
The test in this PR is only checking for `spark.stage.maxConsecutiveAttempts` being below 32k - and I am not sure what is the general behavior expected in the test for #2611 is.

To add, when throwsFetchFailure is enabled, for each stage-attempt, a different celeborn shuffle id is used (all of which map to the same spark shuffle id). The source of confusion could be the use of 'shuffleId' name for the variable - within celeborn it need not map to spark's shuffle id.

Reply:
+CC @ErikFang as well

Reply:
I add a reproducible test
org.apache.celeborn.tests.spark.CelebornFetchFailureSuite#test("celeborn spark integration test - resubmit an unordered barrier stage")
We expect result.size == 10000 and elem._1 == elem._2, but without this PR, we will find that result.size < 10000 and elem._1 is not always equal to elem._2

Reply:
Another option is to change attemptId: Int to attemptId: Long. This way, we can encode (stageAttemptId: Int, attemptId: Int) into attemptId: Long without adding any limitations. WDYT?

Reply:
Thanks for the text, that is really useful @jiang13021 !
So the issue here essentially seems to be that we are not handling barrier mode correctly - essentially, we need something similar to `PbReportShuffleFetchFailure` when a barrier taskset fails and is being reattempted.

Whether it is determinate or indeterminate, it does not matter - barrier stage will be entirely reexecuted always; and so we should essentially throw away its entire output when the stage is being reattempted.

Reply:
> Resubmit stage and reuse celeborn shuffle id
There are at least two valid MapperEnds with same shuffleId, mapId, and taskAttemtNumber

Reply:
The issue is with barrier stages specifically @waitinfuture.
Irrespective of determinism level, spark will rerun all tasks - which is resulting in the issue.

The way to handle it is similar to indeterminate stage - we should discard all output for the barrier stage being re-executed; and so on execution, we will end up generating a new shuffle id.

Reply:
Hi @mridulm ,
> Irrespective of determinism level, spark will rerun all tasks 
IIUC, this statement refers to upstream stage, right? Say task 1 in stage 1 encounters fetch failure, spark will rerun all tasks in the upstream stage 0 because celeborn unregisters all output. When spark returns to stage 1, it will only rerun the failed tasks (in this example task 1), the stage attempt number increases, but the task attempt number of the rerun task1 starts with 0, is this understanding correct? (I'm not 100 percent sure)

Reply:
In the [test case](https://github.com/apache/celeborn/pull/2609/files#diff-e17f15fcca26ddfc412f0af159c784d72417b0f22598e1b1ebfcacd6d4c3ad35R301), it is not a failure due to fetch failure.

The flow would be something like as follows:

Stage 0 completes, generates shuffle data.
Stage 1 (barrier stage) starts, tasks read output from stage 0 - all tasks, except partition 0, complete successfully.
Stage 1, partition 0  attempt 0 throws exception - which results in spark re-running the entire stage 1; since it is a barrier stage.

This means, every other partition in stage 1, which had successfully completed earlier, will also now be reexecuted - and so generate shuffle output for the exact same shuffle id + stage id. Given this is a reexecution of the same stage, and is deterministic, celeborn will reuse the shuffle id even with `throwsFetchFailure` enabled.

[This is a quick prototype](https://github.com/apache/celeborn/compare/main...mridulm:celeborn:fix-barrier-stage-reexecution) to illustrate the fix.

Reply:
Hi @mridulm , I have a question. Forget the test case, if stage 1 is not a barrier stage but a regular one, for stage 1 spark will only re-execute partition 0, the rerun task will still have the same shuffle id + stage id, but different stage attempt id, is this understanding correct?

Reply:
Yes, you are right.

Reply:
OK, so the failed task might have pushed some batches to Celeborn worker before it fails, say Batch0 and Batch1. The re-executed task also pushes its own Batch0 and Batch1 (and more) to Celeborn worker.

For these two batches, the failed task and the re-executed task have the same (mapaId, attemptNumber, batchId), Celeborn client will treat them identical and will filter out either, for example it can filter out re-executed tasks' Batch0 and Batch1, and consume the failed tasks' Batch0 and Batch1.

However, as I said previously, Celeborn's design does not guarantee that re-executed task generates the identical Batches as the failed one (for current implementation, I think Celeborn does generates the same Batches if the task generates the same data, but it's not guaranteed by the design and the behavior might change in the future), so the correct behavior should be that the failed task's Batch0 and Batch1 be filtered out.

When `throwsFetchFailure` is disabled and not barrier stage, this works fine because the re-executed task will not have the same `attemptNumber`. So I think the issue is essentially that stage-rerun can break this design consistency: `Batches with the same (mapId, attemptId, batchId) should contain the identical data`.

Reply:
Yes, agree.
Also note that in the test here specifically even though the RDD is marked determinate, it is not - notice the ordering change when stage is being reattempted (could have been a different set of tues as well).

In general, barrier stages are different from "regular" stages, and need to be handled separately.

Reply:
>IMO we can revisit change to attemptId after we support barrier stage correctly - and evaluate if it is still required. WDYT ?
Agree, as I said in https://issues.apache.org/jira/projects/CELEBORN/issues/CELEBORN-1498?filter=allopenissues, maybe we can just modify the conditions for reusing shuffle id, let DAGScheduler to decide whether to reuse shuffle id.
If DAGScheduler clears appShuffle's numAvailableOutputs, it means it wants celeborn to ignore the previous results, so we generate a new shuffle id. Otherwise, it means it wants to reuse the previous shuffle id and the previous task's output.
There is one additional reason for doing this: RDD#getOutputDeterministicLevel is a protected method of the RDD, which introduces the risk of it being dynamically changed during execution. (Although I've never seen it change). WDYT @mridulm @waitinfuture

Reply:
IMO,  we can specifically consider the situation involving barrier stages, as it is safe to do so. However, considering that Spark is an extensible open-source computing engine, deciding whether to reuse the shuffle ID based on appShuffle's numAvailableOutputs is also a reasonable and sustainable approach.
WDYT @mridulm @waitinfuture @ErikFang @RexXiong

Reply:
Agree with @waitinfuture . In my opinion, not only the barrier stage but also the deterministic stage might lead to data loss if Celeborn doesn't ensure that an attempt task pushes exactly the same data batch.

For a deterministic stage, if the current stage throws a fetch failure exception and the parent stage regenerates the data, this stage will resubmit the failed task with a different stage attempt number but with the same mapId and attemptId. Due to the logic in the LifecycleManager, it will reuse the same Celeborn shuffleId and push data to the same Celeborn partition location. This can cause incorrect data to be stored in Celeborn.

>               val shuffleId: Integer = if (determinate && candidateShuffle.isDefined) {
>                   val id = candidateShuffle.get._1
>                   logInfo(s"reuse existing shuffleId $id for appShuffleId $appShuffleId appShuffleIdentifier $appShuffleIdentifier")
>                  id
>                } else {
>                  val newShuffleId = shuffleIdGenerator.getAndIncrement()
>                  logInfo(s"generate new shuffleId $newShuffleId for appShuffleId $appShuffleId appShuffleIdentifier $appShuffleIdentifier")
>                  shuffleIds.put(appShuffleIdentifier, (newShuffleId, true))
>                  newShuffleId
>                }

For example:

At Stage 1, Stage Attempt 0: use shuffleId 0 generate data (a1, b1, c1, d1)
At Stage 2, Stage Attempt 0: use shuffleId 1 and read partial data from shuffle 0, and push batch0(a1), batch1(b1) to p1, throw fetch failure and disable shuffle 0 
At Stage 1, Stage Attempt 1: use new shuffleId 2 generate data (a1, b1, c1, d1)
At Stage 2, Stage Attempt 1: reuse shuffle 1, and push new batch0(a1, b1), batch1(c1, d1) to same p1, currently celeborn may push those data in different batch in two situation:
- when upstream has multiple partition splits, celeborn will random read those splits, then celeborn would push different batch in two retry.
- if celeborn support spill in future, then batch in celeborn is also different between two retry.

This results in p1 containing map0, attempt0, [batch0(a1), batch1(b1), batch0(a1, b1), batch1(c1, d1)]
At Stage 3, read shuffle 1's data and deduplicate by map0, attempt0 + batchId , then the result will be batch0(a1), batch1(b1), 
this cause data lost(c1, d1)

I reproduced the problem using the following case. I made some modifications to the Celeborn code to throw an exception when data is fetched. Additionally, I implemented random pushing in the SortBasedShuffleWriter:

> val df = sparkSession.createDataFrame(nums, schema)
> val value = df.groupBy("column1").agg(sum("column2").as("sum")).select("column1", "sum")
> Assert.assertEquals(value.count(), 300000)`

The root cause of the issue is that Celeborn currently uses a combination of (mapId, attemptId, batchId) to deduplicate data. However, when a fetch failure occurs, Spark may resubmit the failed tasks with the same stageId but a different stage attempt number, while retaining the same (mapId, attemptNumber). This leads to a scenario where Celeborn cannot distinguish between the results of the two map attempts. If Celeborn does not ensure the shuffle data order for the retried map tasks, the data might become incorrect.

To address this, we need a mechanism to differentiate between the two attempt map tasks, even though Spark uses the same mapId and attempt number. By encoding the stage attempt number with the map attempt number, as  @jiang13021 , we can create a new, unique attempt number for each task. This will allow Celeborn to correctly identify and read only the new successful attempts, ensuring the accuracy of the results.  

If needed, I will organize the code and unit tests I have modified, and we can test them then. 

@jiang13021 @waitinfuture @mridulm

Reply:
Thanks @RexXiong for explanation, please push your test branch and post the link here so that we can review and test.
Hi @mridulm , @RexXiong and I have discussed on this problem for a while and @RexXiong has re-produced the correctness issue for non-barrier stages with minor code change, he will post the code soon, hope this will help making our points more clear.

Reply:
> Thanks @RexXiong for explanation, please push your test branch and post the link here so that we can review and test. Hi @mridulm , @RexXiong and I have discussed on this problem for a while and @RexXiong has re-produced the correctness issue for non-barrier stages with minor code change, he will post the code soon, hope this will help making our points more clear.

You can checkout the [modified code](https://github.com/RexXiong/celeborn/tree/reproduce_deterministic_stage_problem) , and change the value of `spark.celeborn.test.client.push.mockRandomPushForStageRerun` to check the ut `test("celeborn spark integration test - Fetch Failure - Deterministic Stage") ` result. 

And when change the map attempt number use encoding with stageAttemptNumber. the result is correct.

>   public static int getMapAttemptNumber(TaskContext context) {
>    return context.attemptNumber();
>    //    assert (context.stageAttemptNumber() < (1 << 15) && context.attemptNumber() < (1 << 16));
>    //    return (context.stageAttemptNumber() << 16) | context.attemptNumber();
>  }

cc @jiang13021 @mridulm

Reply:
Thanks for the testcase, I will need to look at this in more detail.
One thing I noticed is that the changes to `CelebornInputStream` are not required - that is, even without a task failure, the failure was observed - which is strange.

See [diff here](https://github.com/apache/celeborn/compare/main...mridulm:celeborn:test-shuffle-failure).

Reply:
Ok, the reason for the failure is the change is broken - we are inserting the record twice when `!success || randomPush()`

Let me fix that and try

Reply:
So I fixed this with a generalization of the barrier stage fix - will create a new PR to keep the PR's distinct to help with review.

Reply:
> Ok, the reason for the failure is the change for test is broken - we are inserting the record twice when `!success || randomPush()`
> 
> Let me fix that and try

I see, thanks.

Reply:
I was looking at this more (and again, great test @RexXiong !), and unfortunately it looks like we dont have a better way to fix this than by changing what is `attemptNumber`.
Ideally, I would preferred if this was `taskId`  - but given that would be a breaking protocol (and we would want to backport this to 0.5 imo), it would be practical to do this with the computation that @jiang13021 has proposed here,

Given this, let me revert the changes in #2639 and keep it limited to barrier stage - and current PR (#2609) can fix uniquely identifying a task output.

With this, essentially we have three cases:

* Indeterminate stages - which always throw away shuffle data for parent when a fetch failed is raised (and increments shuffle id).
* Barrier stages - which always throw away shuffle data when current stage encounters any task failure (and increments shuffle id).
* Determinate stage - which will recompute only lost shuffle partitions for parent stage.

In all cases, we ensure that `(shuffleId, mapId, attemptNumber)` is always unique within a spark application (irrespective of recomputations, etc).

Thoughts @waitinfuture, @RexXiong, @jiang13021

Reply:
Hi @mridulm , I have the same opinion with you. I also thought about using `taskId`, in fact I think it's perhaps doable.

The first thing is compatibility, if the protocol is only used in client side then there is no harm because an application always has the same client version during its lifecycle. Currently the server side does read `attemptId` field, but only used for logging like
```
            // TODO just info log for ended attempt
            logWarning(s"Append data failed for task(shuffle $shuffleKey, map $mapId, attempt" +
              s" $attemptId), caused by AlreadyClosedException, endedAttempt $endedAttempt, error message: ${e.getMessage}")
```

Another thing is for Flink (i.e. MapPartition), @RexXiong @FMX can help check is it OK for MapPartition?

Reply:
BTW, encoding stageAttemptId into attemptNumber as this PR does is also accessible for me :)

Reply:
> Hi @mridulm , I have the same opinion with you. I also thought about using `taskId`, in fact I think it's perhaps doable.
> 
> The first thing is compatibility, if the protocol is only used in client side then there is no harm because an application always has the same client version during its lifecycle. Currently the server side does read `attemptId` field, but only used for logging like
> 
> ```
>             // TODO just info log for ended attempt
>             logWarning(s"Append data failed for task(shuffle $shuffleKey, map $mapId, attempt" +
>               s" $attemptId), caused by AlreadyClosedException, endedAttempt $endedAttempt, error message: ${e.getMessage}")
> ```
> 
> Another thing is for Flink (i.e. MapPartition), @RexXiong @FMX can help check is it OK for MapPartition?

After further digging, I found `PartitionFilesSorter` reads `compressedSize` by seeking to 12 bytes offset
```
int compressedSize = Platform.getInt(batchHeader, Platform.BYTE_ARRAY_OFFSET + 12);
```
which prohibits us from changing `attemptId` from `int` to `long`. So considering backward compatibility, seems we can't use `taskI`.

Reply:
Thanks for looking into it @waitinfuture !
Nice analysis.

Given this, agree with you - let us go with the encoded approach

Reply:
@mridulm That's great! Lets moving on this PR :) cc @jiang13021 @RexXiong

Reply:
Resolving this long thread :-)